stdlib/atexit.c

```diff
@@ -15,31 +15,110 @@
 
 #include <stdlib.h>
 #include <string.h>
+#include <sys/threads.h>
 
-void (*atexit_funcs[ATEXIT_MAX])(void);
 
+/* The atexit_node structure uses uint32_t for flags that
+ indicate whether to call a function with arguments or not, therefore
+ why ATEXIT_MAX must be 32 as type "fntype" */
+struct atexit_node {
+	void (*destructors[ATEXIT_MAX])(void);
+	void *args[ATEXIT_MAX];
+	uint32_t fntype;
+	struct atexit_node *prev;
+};
 
-void _atexit_call(void)
+
+/* The first node is statically allocated to provide at least 32 function slots */
+static struct {
+	handle_t lock;
+	struct atexit_node *head;
+} atexit_common = { .head = &((struct atexit_node) {}) };
+
+
+/* Initialise atexit_common structure before main */
+void _atexit_init(void)
+{
+	mutexCreate(&atexit_common.lock);
+	memset(atexit_common.head, 0, sizeof(struct atexit_node));
+}
+
+
+/* Generic function to register destructors */
+static int _atexit_register(int isarg, void (*fn)(void), void *arg)
 {
+	struct atexit_node *node = atexit_common.head;
 	int i;
 
-	for (i = ATEXIT_MAX - 1; i >= 0; i--) {
-		if (atexit_funcs[i] != NULL)
-			atexit_funcs[i]();
+	mutexLock(atexit_common.lock);
+	/* Looking for an empty slot */
+	for (i = 0; i < ATEXIT_MAX; i++) {
+		if (node->destructors[i] == NULL) {
+			break;
+		}
 	}
-}
 
+	/* If there is no free functions slots */
+	if (i == ATEXIT_MAX) {
+		/* to allocate new node */
```

Comment:
Could you merge these two comments into one: allocate new node if there are no free slots left?